tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -108,18 +157,60 @@ public long getFirstShortInProgress() {
     return firstShortInProgress;
   }
 
+  /**
+   * Returns true if the given version corresponds to a transaction that was in-progress at the time this transaction
+   * started.
+   */
   public boolean isInProgress(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0;
   }
 
+  /**
+   * Returns true if the given version is present in one of the arrays of excluded versions (in-progress and
+   * invalid transactions).
+   */
   public boolean isExcluded(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0
       || Arrays.binarySearch(invalids, version) >= 0;
   }
 
+  /**
+   * Returns true if the the given version corresponds to one of the checkpoint versions in the current
+   * transaction.
+   */
+  public boolean isCheckpoint(long version) {
+    return Arrays.binarySearch(checkpointWritePointers, version) >= 0;
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   *
+   * @see #isVisible(long, boolean) to exclude the current write pointer from visible versions.  This method always
+   *      includes the current write pointer.
+   */
   public boolean isVisible(long version) {
+    return isVisible(version, true);
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @param excludeCurrentWritePointer whether writes from the current write pointer should be visible
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   */
+  public boolean isVisible(long version, boolean excludeCurrentWritePointer) {
```

Comment:
@anew You commented on the JIRA:

    An alternative implementation would be to have an option to exclude the current write version
    from a scan. That would simply add the current write version to the exclude filter before 
    submitting the scan request. All other operations would still "read own writes".

As far as I can see, this is exactly what this implementation provides.  You now seem to be arguing for something else.

You also stated:

    What I dislike about the proposed design is that it removes the "read your own writes" 
    capability before and after the checkpoint. Reads that happen after the checkpoint see writes 
    that happened before. But reads before checkpoints don't see prior writes, and reads after 
    checkpoint don't see other writes made after the checkpoint.

    That seems inconsistent with Tephra's normal snapshot isolation.

But this seems to be exactly what you are now proposing.

The goal of this issue to maintain current Tephra semantics (read-your-writes), while providing the knobs necessary to prevent this inducing an infinite loop scenario (the UPSERT/SELECT to same table).  This implementation does this, while only imposing new overhead where it is actually required.

If you would like to introduce alternate isolation levels to Tephra, you are welcome to do so, but that is outside the scope of this issue.

The solution that you are proposing, as far as I understand it, seems to have two problems:

1. When a transaction is started it provides read-your-writes semantics, but as soon as a checkpoint is performed, writes at the current write pointer are no longer visible.  This seems like inconsistent behavior and difficult to reason about since it changes as a side effect of calling the checkpoint operation.
2. As a result, as soon as a checkpoint is performed, a new checkpoint _must_ be performed to see any new writes.  For interleaved write/read operations, this would seem to be a large amount of overhead, where the current solution provides read-your-writes semantics without it.

Phoenix only needs to checkpoint when an UPSERT/SELECT statement is issued against the same table.  At this point it would checkpoint and set the TX_EXCLUDE_CURRENT_WRITE attribute on the Scan operation that it uses for the SELECT portion of that statement.  After this it would not need any special behavior to continue seeing its own writes.  This allows the SELECT portion of the statement to see ALL of the prior writes while just excluding the UPSERTs that are part of the statement.

From @JamesRTaylor's remarks here, it seems like the current implementation would meet their needs.  Again, I do not understand why we would want to build something different.